test/e2e/kappcontroller/package_repo_test.go

```diff
@@ -287,6 +288,414 @@ spec:
+		assert.Contains(t, out, expectedOwnership,
+			"\n========\n Package Ownership Check Failed: expected ", expectedOwnership, "\n=======")
+	}
+
+	appName1 := "pkgrapp1"
+	pkgr1Name := "repo-1.tanzu.carvel.dev"
+
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", appName1})
+	}
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("deploy pkgr1", func() {
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", appName1},
+			e2e.RunOpts{StdinReader: strings.NewReader(pkgr1), AllowError: true})
+		if err != nil {
+			fmt.Println(out)
+			fmt.Println(string(kubectl.Run([]string{"get", "pkgr", "-A", "-oyaml"})))
+		}
+		assert.NoError(t, err)
+
+		out, err = kubectl.RunWithOpts([]string{"get", "packages", "-A"}, e2e.RunOpts{NoNamespace: true})
+		require.Contains(t, out, "shirt-mgr.co.uk.5.5.5")
+		require.Contains(t, out, "shirt-mgr.co.uk.5.6.0")
+		require.Contains(t, out, "coredino.co.uk.32.76.7")
+
+		out = kapp.Run([]string{"inspect", "-a", appName1, "--json"})
+		require.Contains(t, out, `"reconcile_state": "ok"`)
+
+		assertPkgOwnedBy(pkgr1Name, "shirt-mgr.co.uk.5.5.5")
+		assertPkgOwnedBy(pkgr1Name, "shirt-mgr.co.uk.5.6.0")
+		assertPkgOwnedBy(pkgr1Name, "coredino.co.uk.32.76.7")
+	})
+
+	pkgr2 := fmt.Sprintf(
+		"%s%s%s",
+		fmt.Sprintf(pkgrPreamble, 2),
+		fmt.Sprintf(pkgTemplate, "shirt-mgr.co.uk", "5.5.5"),
+		fmt.Sprintf(pkgTemplate, "contooor.co.uk", "0.22.0"))
+	appName2 := "pkgrapp2"
+	pkgr2Name := "repo-2.tanzu.carvel.dev"
+
+	cleanUp2 := func() {
+		kapp.Run([]string{"delete", "-a", appName2})
+	}
+	cleanUp2()
+	defer cleanUp2()
+
+	logger.Section("deploy and check pkgr2", func() {
+		out, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", appName2},
+			e2e.RunOpts{StdinReader: strings.NewReader(pkgr2), AllowError: true})
+		if err != nil {
+			fmt.Println(out)
+			fmt.Println(string(kubectl.Run([]string{"get", "pkgr", "-A", "-oyaml"})))
+		}
+		assert.NoError(t, err)
+
+		out = kapp.Run([]string{"inspect", "-a", appName2, "--json"})
+		require.Contains(t, out, `"reconcile_state": "ok"`)
+
+		assertPkgOwnedBy(pkgr1Name, "shirt-mgr.co.uk.5.5.5")
+		assertPkgOwnedBy(pkgr2Name, "contooor.co.uk.0.22.0")
+	})
+
+	// pkgr3 should fail to reconcile because the contooor package will conflict in the spec
+	pkgr3 := fmt.Sprintf(
+		"%s%s%s",
+		fmt.Sprintf(pkgrPreamble, 3),
+		fmt.Sprintf(pkgTemplate, "shirt-mgr.co.uk", "5.5.5"),
+		strings.Replace(fmt.Sprintf(pkgTemplate, "contooor.co.uk", "0.22.0"), "k8slt/kctrl-example-pkg:v1.0.0", "k8slt/some-other-image:latest", -1))
+	appName3 := "pkgrapp3"
+	pkgr3Name := "repo-3.tanzu.carvel.dev"
+
+	cleanUp3 := func() {
+		kapp.Run([]string{"delete", "-a", appName3})
+	}
+	cleanUp3()
+	defer cleanUp3()
+
+	logger.Section("deploy and check pkgr3", func() {
+		_, err := kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", appName3},
+			e2e.RunOpts{StdinReader: strings.NewReader(pkgr3), AllowError: true})
+		if err != nil {
```

Comment:
shouldnt have conditionals in the test. assert that error is set, and then assert that error contains info.

Reply:
I agree it's a little ugly &verbose -- the trouble is that if it does fail, you have to put this logic back in and run it again to find out why -- all you get is "see status.usefulerrormessage for details" , but once the test has terminated the useful error message is gone. 

should we add support into the e2e.runopts,  like `kubectlOnError: []string{"get", "package" , pkgName, "-oyaml"}` ?